perl-package/AI-MXNet/lib/AI/MXNet/NDArray/Sparse.pm

```diff
@@ -0,0 +1,1342 @@
+            __PACKAGE__->_new_alloc_handle(
+                $self->stype, $self->shape, $other, 1, $self->dtype, $self->_aux_types
+            )
+        );
+        return __PACKAGE__->_copyto($self, out=>$hret)
+    }
+}
+
+=head2 check_format
+
+        Check whether the NDArray format is valid.
+
+        Parameters
+        ----------
+        full_check : bool, optional
+            If `True`, rigorous check, O(N) operations. Otherwise
+            basic check, O(1) operations (default True).
+=cut
+
+method check_format(Bool $full_check=1)
+{
+    scalar(check_call(AI::MXNetCAPI::NDArraySyncCheckFormat($self->handle, $full_check)));
+}
+
+=head2 _data
+
+        A deep copy NDArray of the data array associated with the BaseSparseNDArray.
+
+        This function blocks. Do not use it in performance critical code.
+=cut
+
+method _data()
+{
+    $self->wait_to_read;
+    my $handle = check_call(AI::MXNetCAPI::NDArrayGetDataNDArray($self->handle));
+    return AI::MXNet::NDArray->new(handle => $handle);
+}
+
+=head2 _aux_data
+
+        Get a deep copy NDArray of the i-th aux data array associated with the
+        AI::MXNet::NDArray::Sparse
+
+        This function blocks. Do not use it in performance critical code.
+=cut
+
+method _aux_data(Int $i)
+{
+    $self->wait_to_read;
+    my $handle = check_call(AI::MXNetCAPI::NDArrayGetAuxNDArray($self->handle, $i));
+    return AI::MXNet::NDArray->new(handle => $handle);
+}
+
+package AI::MXNet::NDArray::CSR;
+use AI::MXNet::Base;
+use Mouse;
+extends 'AI::MXNet::NDArray::Sparse';
+
+=head1 NAME
+
+    AI::MXNet::NDArray::CSR - A sparse representation of 2D NDArray in the Compressed Sparse Row format.
+=cut
+
+=head1 DESCRIPTION
+
+    A AI::MXNet::NDArray::CSR represents an AI::MXNet::NDArray as three separate arrays: `data`,
+    `indptr` and `indices`. It uses the CSR representation where the column indices for
+    row i are stored in ``indices[indptr[i]:indptr[i+1]]`` and their corresponding values are stored
+    in ``data[indptr[i]:indptr[i+1]]``.
+
+    The column indices for a given row are expected to be sorted in ascending order.
+    Duplicate column entries for the same row are not allowed.
+
+    Example
+    -------
+    >>> $a = mx->nd->array([[0, 1, 0], [2, 0, 0], [0, 0, 0], [0, 0, 3]]);
+    >>> $a = $a->tostype('csr');
+    >>> $a->data->aspdl;
+    [ 1  2  3]
+    >>> $a->indices->aspdl
+    [1 0 2]
+    >>> $a->indptr->aspdl
+    [0 1 2 2 3]
+
+    See Also
+    --------
+    csr_matrix: Several ways to construct a CSRNDArray
+=cut
+
+#    def __reduce__(self):
+#        return CSRNDArray, (None,), super(CSRNDArray, self).__getstate__()
+
+use overload '+=' => sub { ($_[0] + $_[1])->copyto($_[0]) },
+             '-=' => sub { ($_[0] - $_[1])->copyto($_[0]) },
+             '*=' => sub { ($_[0] * $_[1])->copyto($_[0]) },
+             '/=' => sub { ($_[0] / $_[1])->copyto($_[0]) };
+
+=head2 slice
+
+        Returns a sliced view of this array.
```

Comment:
The documentation is wrong.. Should be a copy instead of a view. I'll post a PR to fix it (in both perl and python).

Reply:
See https://github.com/apache/incubator-mxnet/pull/10034

Reply:
thanks!